cassiopeia/simulator/ClonalSpatialDataSimulator.py

```diff
@@ -0,0 +1,283 @@
+
+    Args:
+        shape: Shape of the space to place cells on. For instance, (100, 100) means a
+            2D surface of 100x100 pixels. By default in two dimensions (length is 2),
+            an elliptical surface is used. In higher dimensions, the entire (hyper)cuboid
+            is used.
+        space: Numpy array mask representing the space that cells may be placed.
+            For example, to place cells on a 2D circlular surface, this argument
+            will be a boolean Numpy array where the circular surface is indicated
+            with True.
+        random_seed: A seed for reproducibility
+
+    Raises:
+        DataSimulatorError if neither `shape` nor `space` are provided
+    """
+    def __init__(
+        self,
+        shape: Optional[Tuple[int, ...]] = None,
+        space: Optional[np.ndarray] = None,
+        random_seed: Optional[int] = None,
+    ):
+        if None in (cv2, disc, neighbors):
+            raise DataSimulatorError(
+                "Some required modules were not found. Make sure you installed Cassiopeia with "
+                "the `spatial` extras, or run `pip install cassiopeia-lineage[spatial]`."
+            )
+
+        if (shape is None) == (space is None):
+            raise DataSimulatorError(
+                "Exactly one of `shape` or `space` must be provided."
+            )
+        self.random_seed = random_seed
+
+        if shape is not None:
+            self.dim = len(shape)
+            if self.dim == 2:
+                center_x = shape[1] // 2
+                center_y = shape[0] // 2
+                self.space = cv2.ellipse(
+                    np.zeros(shape, dtype=np.uint8), (center_x, center_y), (center_x, center_y),
+                    0, 0, 360, 1, -1
+                ).astype(bool)
+            else:
+                self.space = np.ones(shape, dtype=bool)
+
+        if space is not None:
+            self.dim = space.ndim
+            self.space = space
+
+    @staticmethod
+    def __triangulation_graph(points: np.ndarray) -> nx.Graph:
+        """Compute a fully-connected Delaunay triangulation graph from a set of points.
+
+        Args:
+            points: Points to triangulate
+
+        Returns:
+            Networkx graph
+        """
+        tri = spatial.Delaunay(points)
+        G = nx.Graph()
+        for path in tri.simplices:
+            nx.add_path(G, path)
+
+        for n1, n2 in G.edges:
+            G[n1][n2]['weight'] = spatial.distance.euclidean(points[n1], points[n2])
+        return G
+
+    @staticmethod
+    def __nearest_neighbors_graph(points: np.ndarray, k: int) -> nx.Graph:
+        """Compute the nearest neighbors graph from a set of points.
+
+        Args:
+            points: Point coordinates
+            k: Number of nearest neighbors
+
+        Returns:
+            Networkx graph
+        """
+        distances = neighbors.kneighbors_graph(points, k, mode='distance')
+        G = nx.from_scipy_sparse_matrix(distances)
```

Comment:
`from_scipy_sparse_matrix` has been changed to `from_scipy_sparse_array` in Networkx >2.0.0. Needs to be changed before merge

Reply:
It's actually for Networkx>3.0 I believe. The problem is that Networkx 3.0 and greater is only available for python>=3.8. So, I think it's a natural time to deprecate support for 3.7 (we're behind on this anyways) and just focus on >=3.8.